cpp/src/Ice/ObjectAdapterI.cpp

```diff
@@ -1238,49 +1219,65 @@ ObjectAdapterI::computePublishedEndpoints()
         {
-            //
-            // If the PublishedEndpoints property isn't set, we compute the published endpoints
-            // from the OA endpoints, expanding any endpoints that may be listening on INADDR_ANY
-            // to include actual addresses in the published endpoints.
-            //
-            for (unsigned int i = 0; i < _incomingConnectionFactories.size(); ++i)
+            // If the PublishedEndpoints property isn't set, we compute the published endpoints from the factory
+            // endpoints.
+            for (const auto& factory : _incomingConnectionFactories)
+            {
+                endpoints.push_back(factory->endpoint());
+            }
+
+            // Remove all loopback endpoints
+            vector<EndpointIPtr> endpointsNoLoopback;
+            copy_if(
+                endpoints.begin(),
+                endpoints.end(),
+                back_inserter(endpointsNoLoopback),
+                [](const EndpointIPtr& endpoint) { return !endpoint->isLoopback(); });
+
+            // Retrieve published host
+            string publishedHost = _communicator->getProperties()->getProperty(_name + ".PublishedHost");
+
+            if (!endpointsNoLoopback.empty())
             {
-                vector<EndpointIPtr> endps = _incomingConnectionFactories[i]->endpoint()->expandIfWildcard();
-                for (vector<EndpointIPtr>::const_iterator p = endps.begin(); p != endps.end(); ++p)
+                endpoints = endpointsNoLoopback;
+
+                // For non-loopback endpoints, we use the fully qualified name of the local host as default for
+                // publishedHost.
+                if (publishedHost.empty())
+                {
+                    publishedHost = getHostName(); // fully qualified name of local host
+                }
+            }
+            // else keep endpoints as-is; they are all loopback.
+
+            if (!publishedHost.empty())
+            {
+                vector<EndpointIPtr> newEndpoints;
+
+                // Replace the host in all endpoints by publishedHost (when applicable), and remove duplicates.
+                for (const auto& endpoint : endpoints)
                 {
-                    //
-                    // Check for duplicate endpoints, this might occur if an endpoint with a DNS name
-                    // expands to multiple addresses. In this case, multiple incoming connection
-                    // factories can point to the same published endpoint.
-                    //
-                    if (::find(endpoints.begin(), endpoints.end(), *p) == endpoints.end())
+                    EndpointIPtr newEndpoint = endpoint->withPublishedHost(publishedHost);
+                    if (::find_if(
+                            newEndpoints.begin(),
+                            newEndpoints.end(),
+                            [newEndpoint](const EndpointIPtr& p) { return *newEndpoint == *p; }) == newEndpoints.end())
```

Comment:
Should we capture newEndpoint by reference?

Reply:
Updated as suggested.